tests/unit/lms/services/canvas_test.py

```diff
@@ -1,89 +1,282 @@
+        )
+        assert (
+            module_item_configuration.get_canvas_mapped_file_id(sentinel.file_id)
+            == sentinel.found_file_id
+        )
+        assert canvas_api_client.public_url.call_args_list == [
+            call(sentinel.mapped_file_id),
+            call(sentinel.found_file_id),
+        ]
+        assert url == sentinel.url
 
-        assert result == expected_result
-        canvas_api_client.list_files.assert_called_once_with(sentinel.course_id)
+    @pytest.mark.usefixtures("with_mapped_file_id")
+    def test_if_theres_a_permissions_error_and_theres_no_matching_file_it_raises(
+        self, canvas_api_client, canvas_file_finder, public_url_for_file
+    ):
+        canvas_api_client.public_url.side_effect = CanvasAPIPermissionError
+        canvas_file_finder.find_matching_file_in_course.return_value = None
+
+        with pytest.raises(CanvasAPIPermissionError):
+            public_url_for_file(sentinel.file_id)
+
+    def test_it_doesnt_save_a_mapped_file_id_if_getting_that_files_public_url_fails(
+        self,
+        canvas_api_client,
+        canvas_file_finder,
+        module_item_configuration,
+        public_url_for_file,
+    ):
+        canvas_api_client.public_url.side_effect = CanvasAPIPermissionError
+        canvas_file_finder.find_matching_file_in_course.return_value = (
+            sentinel.found_file_id
+        )
+
+        with pytest.raises(CanvasAPIPermissionError):
+            public_url_for_file(sentinel.file_id)
+
+        assert (
+            module_item_configuration.get_canvas_mapped_file_id(sentinel.file_id)
+            == sentinel.file_id
+        )
+
+    @pytest.fixture
+    def module_item_configuration(self, db_session):
+        module_item_configuration = factories.ModuleItemConfiguration()
+        db_session.flush()
+        return module_item_configuration
+
+    @pytest.fixture
+    def with_mapped_file_id(self, module_item_configuration):
+        module_item_configuration.set_canvas_mapped_file_id(
+            sentinel.file_id, sentinel.mapped_file_id
+        )
 
+    @pytest.fixture
+    def public_url_for_file(self, canvas_service, module_item_configuration):
+        return partial(
+            canvas_service.public_url_for_file,
+            module_item_configuration,
+            course_id=sentinel.course_id,
+        )
 
-class TestFindMatchingFileInCourse:
-    def test_it_returns_the_id_if_theres_a_matching_file_in_the_course(
-        self, canvas_service, canvas_api_client
+    @pytest.fixture(autouse=True)
+    def CanvasFileFinder(self, patch):
+        return patch("lms.services.canvas.CanvasFileFinder")
+
+    @pytest.fixture
+    def canvas_file_finder(self, CanvasFileFinder):
+        return CanvasFileFinder.return_value
+
+
+class TestCanvasFileFinder:
+    def test_assert_file_in_course_doesnt_raise_if_the_file_is_in_the_course(
+        self, finder
+    ):
+        finder.assert_file_in_course(sentinel.course_id, "2")
```

Comment:
```suggestion
    def test_assert_file_in_course_doesnt_raise_if_the_file_is_in_the_course(
        self, finder, canvas_api_client
    ):
        canvas_api_client.list_files.return_value = [{"id": 1}, {"id": 2}, {"id": 3}]

        finder.assert_file_in_course(sentinel.course_id, "2")
```

Mentioned below, but this test is a little mysterious on it's own. What's special about "2"?

Reply:
I think this is better now that the fixture has been inlined as suggested:

```python
    def test_assert_file_in_course_doesnt_raise_if_the_file_is_in_the_course(
        self, finder, canvas_api_client
    ):
        canvas_api_client.list_files.return_value = [{"id": sentinel.file_id}]

        finder.assert_file_in_course(sentinel.course_id, str(sentinel.file_id))
```